frontend/src/metabase/lib/data_grid.js

```diff
@@ -1,5 +1,144 @@
 import { formatValue } from "metabase/lib/formatting";
 
+export function multiLevelPivot(
+  data,
+  columnColumnIndexes,
+  rowColumnIndexes,
+  valueColumnIndexes,
+) {
+  const columnColumnValues = [];
+  const rowColumnValues = [];
+  const valueColumnValues = {};
+
+  for (const row of data.rows) {
+    updateValueObject(row, columnColumnIndexes, columnColumnValues);
+    updateValueObject(row, rowColumnIndexes, rowColumnValues);
```

Comment:
Interesting that we call this fn twice here. Might be worth leaving a note about why we need to do that so the flow of the data is easier to follow for someone coming in here for the first time.